File: app/src/main/java/net/nonylene/photolinkviewer/tool/PLVUrlService.java
package net.nonylene.photolinkviewer.tool;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;
import android.preference.PreferenceManager;
import android.widget.Toast;

import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.toolbox.Volley;

import net.nonylene.photolinkviewer.R;
import net.nonylene.photolinkviewer.async.AsyncGetURL;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PLVUrlService {
    private PLVUrlListener plvUrlListener;
    private Context context;

    public PLVUrlService(Context context) {
        this.context = context;
    }

    public interface PLVUrlListener {
        void onGetPLVUrlFinished(PLVUrl plvUrl);
        void onGetPLVUrlFailed(String text);
        void onURLAccepted();
    }

    public void setPLVUrlListener(PLVUrlListener urlListener) {
        plvUrlListener = urlListener;
    }

    public void requestGetPLVUrl(String url) {
        Site site;
        if (url.contains("flickr.com") || url.contains("flic.kr")) {
            site = new FlickrSite(url, context);
        } else if (url.contains("nico.ms") || url.contains("seiga.nicovideo.jp")) {
            site = new NicoSite(url, context);
        } else if (url.contains("pixiv.net")) {
            site = new PixivSite(url, context);
        } else if (url.contains("twimg.com/media/")) {
            site = new TwitterSite(url, context);
        } else if (url.contains("twipple.jp")) {
            site = new TwippleSite(url, context);
        } else if (url.contains("img.ly")) {
            site = new ImglySite(url, context);
        } else if (url.contains("instagram.com") || url.contains("instagr.am")) {
            site = new InstagramSite(url, context);
        } else if (url.contains("gyazo.com")) {
            site = new GyazoSite(url, context);
        } else if (url.contains("imgur.com")) {
            site = new ImgurSite(url, context);
        } else {
            site = new OtherSite(url, context);
        }
        site.setPLVUrlListener(plvUrlListener);
        site.getPLVUrl();
    }

    private class Site {
        protected String url;
        protected PLVUrlListener listener;
        protected Context context;

        public Site(String url, Context context) {
            this.url = url;
            this.context = context;
        }

        public void setPLVUrlListener(PLVUrlListener listener) {
            this.listener = listener;
        }

        protected boolean wifiChecker(SharedPreferences sharedPreferences) {
            //check wifi connecting and setting or not
            boolean wifi = false;
            if (sharedPreferences.getBoolean("wifi_switch", false)) {
                // get wifi status
                ConnectivityManager manager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
                NetworkInfo info = manager.getActiveNetworkInfo();
                if (info.getType() == ConnectivityManager.TYPE_WIFI) {
                    wifi = true;
                }
            }
            return wifi;
        }

        public void getPLVUrl() {
        }

        protected String getQuality(String siteName) {
            SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);

            String quality;
            if (wifiChecker(sharedPreferences)) {
                quality = sharedPreferences.getString(siteName + "_quality_wifi", "large");
            } else {
                quality = sharedPreferences.getString(siteName + "_quality_3g", "large");
            }

            return quality;
        }

        protected void onParseFailed() {
            listener.onGetPLVUrlFailed(context.getString(R.string.url_purse_toast));
        }
    }

    private class TwitterSite extends Site {

        public TwitterSite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            PLVUrl plvUrl = new PLVUrl(url);
            Pattern pattern = Pattern.compile("^https?://pbs\\.twimg\\.com/media/([^\\.]+)\\.");
            Matcher matcher = pattern.matcher(url);
            if (!matcher.find()) {
                super.onParseFailed();
                return;
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("twitter");

            String id = matcher.group(1);
            plvUrl.setFileName(id);

            String file_url = null;
            switch (super.getQuality("twitter")) {
                case "original":
                    file_url = url + ":orig";
                    break;
                case "large":
                    file_url = url + ":large";
                    break;
                case "medium":
                    file_url = url;
                    break;
                case "small":
                    file_url = url + ":small";
                    break;
            }
            plvUrl.setDisplayUrl(file_url);
            plvUrl.setThumbUrl(url + ":small");
            plvUrl.setBiggestUrl(url + ":orig");

            listener.onGetPLVUrlFinished(plvUrl);
        }
    }

    private class TwippleSite extends Site {

        public TwippleSite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            PLVUrl plvUrl = new PLVUrl(url);
            Pattern pattern = Pattern.compile("^https?://p\\.twipple\\.jp/(\\w+)");
            Matcher matcher = pattern.matcher(url);
            if (!matcher.find()) {
                super.onParseFailed();
                return;
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("twipple");

            String id = matcher.group(1);
            plvUrl.setFileName(id);

            String file_url = null;
            switch (super.getQuality("twipple")) {
                case "original":
                    file_url = "http://p.twipple.jp/show/orig/" + id;
                    break;
                case "large":
                    file_url = "http://p.twipple.jp/show/large/" + id;
                    break;
                case "thumb":
                    file_url = "http://p.twipple.jp/show/thumb/" + id;
                    break;
            }
            plvUrl.setDisplayUrl(file_url);
            plvUrl.setThumbUrl("http://p.twipple.jp/show/large/" + id);
            plvUrl.setBiggestUrl("http://p.twipple.jp/show/orig/" + id);

            listener.onGetPLVUrlFinished(plvUrl);
        }
    }

    private class ImglySite extends Site {

        public ImglySite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            PLVUrl plvUrl = new PLVUrl(url);
            Pattern pattern = Pattern.compile("^https?://img\\.ly/(\\w+)");
            Matcher matcher = pattern.matcher(url);
            if (!matcher.find()) {
                super.onParseFailed();
                return;
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("img.ly");

            String id = matcher.group(1);
            plvUrl.setFileName(id);

            String file_url = null;
            switch (super.getQuality("imgly")) {
                case "full":
                    file_url = "http://img.ly/show/full/" + id;
                    break;
                case "large":
                    file_url = "http://img.ly/show/large/" + id;
                    break;
                case "medium":
                    file_url = "http://img.ly/show/medium/" + id;
                    break;
            }
            plvUrl.setDisplayUrl(file_url);
            plvUrl.setThumbUrl("http://img.ly/show/medium/" + id);
            plvUrl.setBiggestUrl("http://img.ly/show/full/" + id);

            listener.onGetPLVUrlFinished(plvUrl);
        }
    }

    private class InstagramSite extends Site {

        public InstagramSite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            PLVUrl plvUrl = new PLVUrl(url);
            Pattern pattern = Pattern.compile("^https?://instagr\\.?am[\\.com]*/p/([^/\\?=]+)");
            Matcher matcher = pattern.matcher(url);
            if (!matcher.find()) {
                super.onParseFailed();
                return;
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("instagram");

            String id = matcher.group(1);
            plvUrl.setFileName(id);

            String file_url = null;
            switch (super.getQuality("instagram")) {
                case "large":
                    file_url = "https://instagram.com/p/" + id + "/media/?size=l";
                    break;
                case "medium":
                    file_url = "https://instagram.com/p/" + id + "/media/?size=m";
                    break;
            }
            plvUrl.setDisplayUrl(file_url);
            plvUrl.setThumbUrl("https://instagram.com/p/" + id + "/media/?size=m");
            plvUrl.setBiggestUrl("https://instagram.com/p/" + id + "/media/?size=l");

            listener.onGetPLVUrlFinished(plvUrl);
        }
    }

    private class GyazoSite extends Site {

        public GyazoSite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            PLVUrl plvUrl = new PLVUrl(url);
            Pattern pattern = Pattern.compile("^https?://.*gyazo\\.com/(\\w+)");
            Matcher matcher = pattern.matcher(url);
            if (!matcher.find()) {
                super.onParseFailed();
                return;
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("gyazo");

            String id = matcher.group(1);
            plvUrl.setFileName(id);

            String file_url = "https://gyazo.com/" + id + "/raw";
            plvUrl.setDisplayUrl(file_url);

            listener.onGetPLVUrlFinished(plvUrl);
        }
    }

    private class ImgurSite extends Site {

        public ImgurSite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            PLVUrl plvUrl = new PLVUrl(url);
            Pattern pattern = Pattern.compile("^https?://.*imgur\\.com/([\\w^\\.]+)");
            Matcher matcher = pattern.matcher(url);
            if (!matcher.find()) {
                super.onParseFailed();
                return;
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("imgur");

            String id = matcher.group(1);
            plvUrl.setFileName(id);

            String file_url = "http://i.imgur.com/" + id + ".jpg";
            plvUrl.setDisplayUrl(file_url);

            listener.onGetPLVUrlFinished(plvUrl);
        }
    }

    private class OtherSite extends Site {

        public OtherSite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            PLVUrl plvUrl = new PLVUrl(url);
            Pattern pattern = Pattern.compile("/([^\\./]+)\\.?[\\w\\?=]*$");
            Matcher matcher = pattern.matcher(url);
            if (!matcher.find()) {
                super.onParseFailed();
                return;
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("other");

            String id = matcher.group(1);
            plvUrl.setFileName(id);

            plvUrl.setDisplayUrl(url);

            listener.onGetPLVUrlFinished(plvUrl);
        }
    }

    private class PixivSite extends Site {

        public PixivSite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            final PLVUrl plvUrl = new PLVUrl(url);
            Pattern pattern = Pattern.compile("^https?://.*pixiv\\.net/member_illust.php?.*illust_id=(\\d+)");
            Matcher matcher = pattern.matcher(url);
            if (!matcher.find()) {
                super.onParseFailed();
                return;
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("pixiv");

            String id = matcher.group(1);
            plvUrl.setFileName(id);

            RequestQueue queue = Volley.newRequestQueue(context);
            queue.add(new PXVStringRequest(context, id,
                    new Response.Listener<String>() {
                        @Override
                        public void onResponse(String response) {
                            final String[] list = response.split(",", -1);
                            // get original photo
                            try {
                                String file_url = list[9].replaceAll("\"", "");

                                plvUrl.setDisplayUrl(file_url);

                                listener.onGetPLVUrlFinished(plvUrl);

                            } catch (ArrayIndexOutOfBoundsException e) {
                                listener.onGetPLVUrlFailed("Cannot open. R-18?");
                            }
                        }
                    }));
        }
    }

    private class FlickrSite extends Site {

        public FlickrSite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            final PLVUrl plvUrl = new PLVUrl(url);

            String id = null;
            if (url.contains("flickr")) {
                Pattern pattern = Pattern.compile("^https?://[wm]w*\\.flickr\\.com/?#?/photos/[\\w@]+/(\\d+)");
                Matcher matcher = pattern.matcher(url);
                if (!matcher.find()) {
                    super.onParseFailed();
                    return;
                }
                id = matcher.group(1);
            } else if (url.contains("flic.kr")) {
                Pattern pattern = Pattern.compile("^https?://flic\\.kr/p/(\\w+)");
                Matcher matcher = pattern.matcher(url);
                if (!matcher.find()) {
                    super.onParseFailed();
                    return;
                }
                id = Base58.decode(matcher.group(1));
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("flickr");

            plvUrl.setFileName(id);

            String api_key = (String) context.getText(R.string.flickr_key);
            String request = "https://api.flickr.com/services/rest/?method=flickr.photos.getInfo&format=json&api_key=" + api_key +
                    "&photo_id=" + id;

            RequestQueue queue = Volley.newRequestQueue(context);
            queue.add(new MyJsonObjectRequest(context, request,
                    new Response.Listener<JSONObject>() {
                        @Override
                        public void onResponse(JSONObject response) {
                            try {
                                listener.onGetPLVUrlFinished(parseFlickr(response, plvUrl));
                            } catch (JSONException e) {
                                listener.onGetPLVUrlFailed(context.getString(R.string.show_flickrjson_toast));
                                e.printStackTrace();
                            }
                        }
                    }
            ));
        }

        private PLVUrl parseFlickr(JSONObject json, PLVUrl plvUrl) throws JSONException {
            //for flickr
            JSONObject photo = new JSONObject(json.getString("photo"));
            String farm = photo.getString("farm");
            String server = photo.getString("server");
            String id = photo.getString("id");
            String secret = photo.getString("secret");

            String file_url = null;
            switch (super.getQuality("flickr")) {
                case "original":
                    String original_secret = photo.getString("originalsecret");
                    String original_format = photo.getString("originalformat");
                    file_url = "https://farm" + farm + ".staticflickr.com/" + server + "/" + id + "_" + original_secret + "_o." + original_format;
                    break;
                case "large":
                    file_url = "https://farm" + farm + ".staticflickr.com/" + server + "/" + id + "_" + secret + "_b.jpg";
                    break;
                case "medium":
                    file_url = "https://farm" + farm + ".staticflickr.com/" + server + "/" + id + "_" + secret + "_z.jpg";
                    break;
            }


            String original_secrets = photo.getString("originalsecret");
            String original_formats = photo.getString("originalformat");
            String biggestUrl = "https://farm" + farm + ".staticflickr.com/" + server + "/" + id + "_" + original_secrets + "_o." + original_formats;

            plvUrl.setDisplayUrl(file_url);
            plvUrl.setThumbUrl("https://farm" + farm + ".staticflickr.com/" + server + "/" + id + "_" + secret + "_z.jpg");
            plvUrl.setBiggestUrl(biggestUrl);

            return plvUrl;
        }
    }

    private class NicoSite extends Site {

        public NicoSite(String url, Context context) {
            super(url, context);
        }

        @Override
        public void getPLVUrl() {
            super.getPLVUrl();

            final PLVUrl plvUrl = new PLVUrl(url);

            Pattern pattern;
            if (url.contains("nico.ms")) {
                pattern = Pattern.compile("^https?://nico\\.ms/im(\\d+)");
            } else {
                pattern = Pattern.compile("^https?://seiga.nicovideo.jp/seiga/im(\\d+)");
            }
            Matcher matcher = pattern.matcher(url);
            if (!matcher.find()) {
                super.onParseFailed();
                return;
            }

            listener.onURLAccepted();

            plvUrl.setSiteName("nico");

            final String id = matcher.group(1);
            plvUrl.setFileName(id);

            String oldUrl = "http://seiga.nicovideo.jp/image/source/" + id;

            AsyncTask<String, Integer, String> task = new AsyncGetURL() {
                @Override
                protected void onPostExecute(String redirect) {
                    super.onPostExecute(redirect);
                    listener.onGetPLVUrlFinished(parseNico(redirect, id, plvUrl));
                }
            };

            task.execute(oldUrl);
        }

        private PLVUrl parseNico(String redirect, String id, PLVUrl plvUrl) {
            String biggest_url = redirect.replace("/o/", "/priv/");

            SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);

            boolean original;
            if (wifiChecker(sharedPreferences)) {
                original = sharedPreferences.getBoolean("original_switch_wifi", false);
            } else {
                original = sharedPreferences.getBoolean("original_switch_3g", false);
            }

            String quality = super.getQuality("nicoseiga");

            String file_url = null;
            switch (quality) {
                case "original":
                    file_url = biggest_url;
                    break;
                case "large":
                    file_url = "http://lohas.nicoseiga.jp/img/" + id + "l";
                    break;
                case "medium":
                    file_url = "http://lohas.nicoseiga.jp/img/" + id + "m";
                    break;
            }

            if (redirect.contains("account.nicovideo.jp") && (original || quality.equals("original"))) {
                // cannot preview original photo
                biggest_url = "http://lohas.nicoseiga.jp/img/" + id + "l";
                Toast.makeText(context, context.getString(R.string.nico_original_toast), Toast.LENGTH_LONG).show();
            }

            plvUrl.setDisplayUrl(file_url);
            plvUrl.setThumbUrl("http://lohas.nicoseiga.jp/img/" + id + "m");
            plvUrl.setBiggestUrl(biggest_url);

            return plvUrl;
        }
    }
}

File: app/src/main/java/net/nonylene/photolinkviewer/tool/PLVUrlService.kt
package net.nonylene.photolinkviewer.tool

import android.content.Context
import android.content.SharedPreferences
import android.net.ConnectivityManager
import android.preference.PreferenceManager
import android.util.Base64
import android.widget.Toast

import com.android.volley.Response
import com.android.volley.toolbox.Volley

import net.nonylene.photolinkviewer.R
import net.nonylene.photolinkviewer.async.AsyncGetURL

import org.json.JSONException
import org.json.JSONObject

import java.util.regex.Pattern

import javax.crypto.spec.SecretKeySpec

class PLVUrlService(private val context: Context, private val plvUrlListener: PLVUrlService.PLVUrlListener) {

    interface PLVUrlListener {
        fun onGetPLVUrlFinished(plvUrls: Array<PLVUrl>)
        fun onGetPLVUrlFailed(text: String)
        fun onURLAccepted()
    }

    fun requestGetPLVUrl(url: String) {
        val site = when {
            url.contains("flickr.com"), url.contains("flic.kr")
                                             -> FlickrSite(url, context, plvUrlListener)
            url.contains("nico.ms"), url.contains("seiga.nicovideo.jp")
                                             -> NicoSite(url, context, plvUrlListener)
            url.contains("twimg.com/media/") -> TwitterSite(url, context, plvUrlListener)
            url.contains("twipple.jp")       -> TwippleSite(url, context, plvUrlListener)
            url.contains("img.ly")           -> ImglySite(url, context, plvUrlListener)
            url.contains("instagram.com"), url.contains("instagr.am")
                                             -> InstagramSite(url, context, plvUrlListener)
            url.contains("gyazo.com")        -> GyazoSite(url, context, plvUrlListener)
            url.contains("imgur.com")        -> ImgurSite(url, context, plvUrlListener)
            url.contains("vine.co")          -> VineSite(url, context, plvUrlListener)
            else                             -> OtherSite(url, context, plvUrlListener)
        }
        site.getPLVUrl()
    }

    private abstract inner class Site(protected var url: String, protected var context: Context, protected val listener: PLVUrlListener) {

        protected fun wifiChecker(sharedPreferences: SharedPreferences): Boolean {
            //check wifi connecting and wifi setting enabled or not
            return sharedPreferences.getBoolean("wifi_switch", false) &&
                    (context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager).activeNetworkInfo.type ==
                            ConnectivityManager.TYPE_WIFI
        }

        protected fun getQuality(siteName: String): String {
            val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
            return sharedPreferences.getString("${siteName}_quality_" +
                    if (wifiChecker(sharedPreferences)) "wifi" else "3g", "large")
        }

        protected fun getId(url: String, regex: String): String? {
            val matcher = Pattern.compile(regex).matcher(url)
            if (!matcher.find()) {
                onParseFailed()
                return null
            }
            listener.onURLAccepted()
            return matcher.group(1)
        }

        protected fun onParseFailed() {
            listener.onGetPLVUrlFailed(context.getString(R.string.url_purse_toast))
        }

        abstract fun getPLVUrl()
    }

    private inner class TwitterSite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            super.getId(url, "^https?://pbs\\.twimg\\.com/media/([^\\.]+)\\.")?.let { id ->
                val plvUrl = PLVUrl(url)

                plvUrl.siteName = "twitter"
                plvUrl.fileName = id

                plvUrl.biggestUrl = url + ":orig"
                plvUrl.thumbUrl = url + ":small"
                plvUrl.displayUrl = when (super.getQuality("twitter")) {
                    "original" -> plvUrl.biggestUrl
                    "large"    -> url + ":large"
                    "medium"   -> url
                    "small"    -> plvUrl.thumbUrl
                    else       -> null
                }

                listener.onGetPLVUrlFinished(arrayOf(plvUrl))
            }
        }
    }

    private inner class TwippleSite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            super.getId(url, "^https?://p\\.twipple\\.jp/(\\w+)")?.let { id ->
                val plvUrl = PLVUrl(url)
                plvUrl.siteName = "twipple"
                plvUrl.fileName = id

                plvUrl.biggestUrl = "http://p.twipple.jp/show/orig/" + id
                plvUrl.thumbUrl = "http://p.twipple.jp/show/large/" + id
                plvUrl.displayUrl = when (super.getQuality("twipple")) {
                    "original" -> plvUrl.biggestUrl
                    "large"    -> plvUrl.thumbUrl
                    "thumb"    -> "http://p.twipple.jp/show/thumb/" + id
                    else       -> null
                }

                listener.onGetPLVUrlFinished(arrayOf(plvUrl))
            }
        }
    }

    private inner class ImglySite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            super.getId(url, "^https?://img\\.ly/(\\w+)")?.let { id ->
                val plvUrl = PLVUrl(url)

                plvUrl.siteName = "imgly"
                plvUrl.fileName = id

                plvUrl.biggestUrl = "http://img.ly/show/full/" + id
                plvUrl.thumbUrl = "http://img.ly/show/medium/" + id
                plvUrl.displayUrl = when (super.getQuality("imgly")) {
                    "full"   -> plvUrl.biggestUrl
                    "large"  -> plvUrl.thumbUrl
                    "medium" -> "http://img.ly/show/medium/" + id
                    else     -> null
                }

                listener.onGetPLVUrlFinished(arrayOf(plvUrl))
            }
        }
    }

    private inner class InstagramSite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            super.getId(url, "^https?://instagr\\.?am[\\.com]*/p/([^/\\?=]+)")?.let { id ->
                val plvUrl = PLVUrl(url)

                plvUrl.siteName = "instagram"
                plvUrl.fileName = id

                if (PreferenceManager.getDefaultSharedPreferences(context).getBoolean("instagram_api", false)) {

                    val preferences = context.getSharedPreferences("preference", Context.MODE_PRIVATE)

                    if (!preferences.getBoolean("instagram_authorized", false)) {
                        listener.onGetPLVUrlFailed("You have to authorize instagram account or change instagram api preference.")
                        return
                    }

                    val keyByte = Base64.decode(preferences.getString("instagram_key", null), Base64.DEFAULT)
                    val tokenByte = Base64.decode(preferences.getString("instagram_token", null), Base64.DEFAULT)
                    val token = Encryption.decrypt(tokenByte, SecretKeySpec(keyByte, 0, keyByte.size(), "AES"))
                    val apiUrl = "https://api.instagram.com/v1/media/shortcode/${id}?access_token=${token}"

                    Volley.newRequestQueue(context).add(MyJsonObjectRequest(context, apiUrl,
                            Response.Listener { response ->
                                try {
                                    listener.onGetPLVUrlFinished(arrayOf(parseInstagram(response, plvUrl)))
                                } catch (e: JSONException) {
                                    listener.onGetPLVUrlFailed("instagram JSON Parse Error!")
                                    e.printStackTrace()
                                }
                            })
                    );

                } else {
                    plvUrl.biggestUrl = "https://instagram.com/p/${id}/media/?size=l"
                    plvUrl.displayUrl = when (super.getQuality("instagram")) {
                        "large"  -> plvUrl.biggestUrl
                        "medium" -> "https://instagram.com/p/${id}/media/?size=m"
                        else     -> null
                    }
                    plvUrl.thumbUrl = "https://instagram.com/p/${id}/media/?size=m"

                    listener.onGetPLVUrlFinished(arrayOf(plvUrl))
                }

            }
        }

        @Throws(JSONException::class)
        private fun parseInstagram(json: JSONObject, plvUrl: PLVUrl): PLVUrl {
            //for flickr
            val data = JSONObject(json.getString("data"))
            val fileUrls: JSONObject

            if ("video" == data.getString("type")) {
                plvUrl.setIsVideo(true)
                fileUrls = data.getJSONObject("videos")
            } else {
                fileUrls = data.getJSONObject("images")
            }

            plvUrl.displayUrl =
                    when (super.getQuality("instagram")) {
                        "large"  -> fileUrls.getJSONObject("standard_resolution")
                        "medium" -> fileUrls.getJSONObject("low_resolution")
                        else     -> null
                    }!!.getString("url")

            val imageUrls = data.getJSONObject("images")
            plvUrl.thumbUrl = imageUrls.getJSONObject("low_resolution").getString("url")
            plvUrl.biggestUrl = imageUrls.getJSONObject("standard_resolution").getString("url")

            return plvUrl
        }
    }

    private inner class GyazoSite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            super.getId(url, "^https?://.*gyazo\\.com/(\\w+)")?.let { id ->
                val plvUrl = PLVUrl(url)

                plvUrl.siteName = "gyazo"
                plvUrl.fileName = id

                plvUrl.displayUrl = "https://gyazo.com/${id}/raw"

                listener.onGetPLVUrlFinished(arrayOf(plvUrl))
            }
        }
    }

    private inner class ImgurSite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            super.getId(url, "^https?://.*imgur\\.com/([\\w^\\.]+)")?.let { id ->
                val plvUrl = PLVUrl(url)

                plvUrl.siteName = "imgur"
                plvUrl.fileName = id

                val file_url = "http://i.imgur.com/${id}.jpg"
                plvUrl.displayUrl = file_url

                listener.onGetPLVUrlFinished(arrayOf(plvUrl))
            }
        }
    }

    private inner class OtherSite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            super.getId(url, "/([^\\./]+)\\.(png|jpg|jpeg|gif)[\\w\\?=]*$")?.let { id ->
                val plvUrl = PLVUrl(url)

                plvUrl.siteName = "other"
                plvUrl.fileName = id

                plvUrl.displayUrl = url

                listener.onGetPLVUrlFinished(arrayOf(plvUrl))
            }
        }
    }

    private inner class FlickrSite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            when {
                url.contains("flickr")  -> super.getId(url, "^https?://[wm]w*\\.flickr\\.com/?#?/photos/[\\w@]+/(\\d+)")
                url.contains("flic.kr") -> super.getId(url, "^https?://flic\\.kr/p/(\\w+)")?.let { Base58.decode(it) }
                else                    -> null
            }?.let { id ->
                val plvUrl = PLVUrl(url)

                plvUrl.siteName = "flickr"
                plvUrl.fileName = id

                val api_key = context.getText(R.string.flickr_key) as String
                val request = "https://api.flickr.com/services/rest/?method=flickr.photos.getInfo&format=json&nojsoncallback=1&api_key=${api_key}&photo_id=${id}"

                Volley.newRequestQueue(context).add(MyJsonObjectRequest(context, request,
                        Response.Listener { response ->
                            try {
                                listener.onGetPLVUrlFinished(arrayOf(parseFlickr(response, plvUrl)))
                            } catch (e: JSONException) {
                                listener.onGetPLVUrlFailed(context.getString(R.string.show_flickrjson_toast))
                                e.printStackTrace()
                            }
                        })
                );
            }
        }

        @Throws(JSONException::class)
        private fun parseFlickr(json: JSONObject, plvUrl: PLVUrl): PLVUrl {
            //for flickr
            val photo = JSONObject(json.getString("photo"))
            val farm = photo.getString("farm")
            val server = photo.getString("server")
            val id = photo.getString("id")
            val secret = photo.getString("secret")

            val original_secrets = photo.getString("originalsecret")
            val original_formats = photo.getString("originalformat")

            plvUrl.biggestUrl = "https://farm" + farm + ".staticflickr.com/" + server + "/" + id + "_" + original_secrets + "_o." + original_formats
            plvUrl.thumbUrl = "https://farm" + farm + ".staticflickr.com/" + server + "/" + id + "_" + secret + "_z.jpg"
            plvUrl.displayUrl = when (super.getQuality("flickr")) {
                "original" -> plvUrl.biggestUrl
                "large"    -> "https://farm${farm}.staticflickr.com/${server}/${id}_${secret}_b.jpg"
                "medium"   -> plvUrl.thumbUrl
                else       -> null
            }

            return plvUrl
        }
    }

    private inner class NicoSite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            when {
                url.contains("nico.ms") -> super.getId(url, "^https?://nico\\.ms/im(\\d+)")
                else                    -> super.getId(url, "^https?://seiga.nicovideo.jp/seiga/im(\\d+)")
            }?.let { id ->
                val plvUrl = PLVUrl(url)

                plvUrl.siteName = "nico"
                plvUrl.fileName = id

                val task = object : AsyncGetURL() {
                    override fun onPostExecute(redirect: String) {
                        super.onPostExecute(redirect)
                        listener.onGetPLVUrlFinished(arrayOf(parseNico(redirect, id, plvUrl)))
                    }
                }

                task.execute("http://seiga.nicovideo.jp/image/source/" + id)
            }
        }

        private fun parseNico(redirect: String, id: String, plvUrl: PLVUrl): PLVUrl {
            var biggest_url = redirect.replace("/o/", "/priv/")

            val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)

            val original: Boolean
            if (wifiChecker(sharedPreferences)) {
                original = sharedPreferences.getBoolean("original_switch_wifi", false)
            } else {
                original = sharedPreferences.getBoolean("original_switch_3g", false)
            }

            val quality = super.getQuality("nicoseiga")

            if (redirect.contains("account.nicovideo.jp") && (original || quality == "original")) {
                // cannot preview original photo
                biggest_url = "http://lohas.nicoseiga.jp/img/" + id + "l"
                Toast.makeText(context, context.getString(R.string.nico_original_toast), Toast.LENGTH_LONG).show()
            }

            plvUrl.biggestUrl = biggest_url
            plvUrl.thumbUrl = "http://lohas.nicoseiga.jp/img/" + id + "m"
            plvUrl.displayUrl = when (quality) {
                "original" -> biggest_url
                "large"    -> "http://lohas.nicoseiga.jp/img/" + id + "l"
                "medium"   -> plvUrl.thumbUrl
                else       -> null
            }

            return plvUrl
        }
    }

    private inner class VineSite(url: String, context: Context, listener: PLVUrlListener) : Site(url, context, listener) {

        override fun getPLVUrl() {
            super.getId(url, "^https?://vine\\.co/v/(\\w+)")?.let { id ->
                val plvUrl = PLVUrl(url)

                plvUrl.siteName = "vine"
                plvUrl.setIsVideo(true)
                plvUrl.fileName = id

                val request = "https://api.vineapp.com/timelines/posts/s/" + id

                Volley.newRequestQueue(context).add(MyJsonObjectRequest(context, request,
                        Response.Listener { response ->
                            try {
                                listener.onGetPLVUrlFinished(arrayOf(parseVine(response, plvUrl)))
                            } catch (e: JSONException) {
                                listener.onGetPLVUrlFailed(context.getString(R.string.show_flickrjson_toast))
                                e.printStackTrace()
                            }
                        })
                );
            }
        }

        @Throws(JSONException::class)
        private fun parseVine(json: JSONObject, plvUrl: PLVUrl): PLVUrl {
            val records = json.getJSONObject("data").getJSONArray("records").getJSONObject(0)
            plvUrl.displayUrl = records.getString("videoUrl")
            plvUrl.thumbUrl = records.getString("thumbnailUrl")
            return plvUrl
        }
    }
}
